### PR TITLE
Remove is_spotlight field from `Trait`

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Small PR, only the last commit is relevant here. The rest is coming from #80883 because I need the `TyCtxt` stored inside `Cache`.

The point is to make ItemKind looks as close as possible to the compiler type so that it makes the switch simpler (which is why I make all these "small" PRs).

r? @jyn514

### CODE REVIEW

File: src/librustdoc/clean/utils.rs
Comment (LingMan) [inline src/librustdoc/clean/utils.rs:469]: Would be a bit shorter as:
```suggestion
    attrs.iter().any(|attr| {
        attr.has_name(sym::doc)
            && attr.meta_item_list().map_or(false, |l| list_contains_name(&l, flag))
    })
```

Comment (GuillaumeGomez) [reply src/librustdoc/clean/utils.rs:469]: Indeed!

**Comment by jyn514** [comment]
@camelid would you like to review?

File: src/librustdoc/clean/inline.rs
Comment (camelid) [inline src/librustdoc/clean/inline.rs:192]: We should probably rename this flag to `notable_trait` since (a) it's clearer and (b) that's what the popup that it triggers is called. We should probably do it in another PR though; I'll try to do that soon.

**Comment by camelid** [comment]
@jyn514 Thanks for giving me the opportunity to review! :)

I'll give it a look, but it's probably still a good idea for you to take a look as well since I'm still learning the intricacies of rustdoc.

**Comment by camelid** [comment]
(Marking as blocked on #80883.)

File: src/librustdoc/clean/utils.rs
Comment (camelid) [inline src/librustdoc/clean/utils.rs:465]: Why is this being added? Is it different from [`Attributes::has_doc_flag`](https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/clean/types/struct.Attributes.html#method.has_doc_flag)? If so, that should probably be documented so we don't confuse people.

File: src/librustdoc/html/render/mod.rs
Comment (camelid) [inline src/librustdoc/html/render/mod.rs:3715]: As Joshua said in #80883, `TyCtxt` is Copy so you don't need `as_ref()`:

```suggestion
                    clean::utils::has_doc_flag(c.tcx.unwrap().get_attrs(d), sym::spotlight)
```

**Review by camelid — COMMENTED** [review]
Is this field being removed because it's only used in one place and we don't want to pile stuff onto `clean::Trait`?

File: src/librustdoc/html/render/mod.rs
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:3715]: Indeed!

File: src/librustdoc/clean/utils.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/utils.rs:465]: One takes the `hir::Attributes` whereas the other takes `clean::Attributes`. I'll add a doc comment.

File: src/librustdoc/clean/inline.rs
Comment (camelid) [reply src/librustdoc/clean/inline.rs:192]: Opened #80936.

File: src/librustdoc/clean/utils.rs
Comment (camelid) [reply src/librustdoc/clean/utils.rs:465]: Huh, I think I see. It's unfortunate to have them duplicated :/

**Comment by GuillaumeGomez** [comment]
Rebased! Only the last commit is from this PR, the others still come from #80883.

cc @CraftSpider @ollie27 
r? @camelid

**Review by CraftSpider — COMMENTED** [review]
Looks good. As far as I can tell this should behave the same, just with spotlight calculated at use instead of early. Only question is: How is spotlight_decl called, will this end up duplicating the spotlight lookup? Not too terrible if it does, but want to make sure I understand.

**Comment by GuillaumeGomez** [comment]
That's actually a pretty good point. It forces us to run this check every time a function returns a "spotlighted" item. That's not great.

**Comment by CraftSpider** [comment]
I wonder if this is the kind of thing that could be stored in the cache? Not sure if the cache is *just* for DefId stuff, and that would cause more issues, but it would allow removing it from the clean AST and still only calculating it as needed

**Comment by GuillaumeGomez** [comment]
It's kinda the case now. The idea here was to reduce the memory usage and maybe improve the performance too, but not sure if it's the right approach...

**Comment by camelid** [comment]
> The idea here was to reduce the memory usage [...]

`bool` is only one byte, so I wouldn't think this would reduce memory usage *that* much. Just throwing this out there: would this be something to store in `DocContext`? E.g., we'd have a field like:

```rust
// or `notable_traits` once my PR lands
spotlighted_traits: HashMap<DefId, bool>
```

**Comment by GuillaumeGomez** [comment]
@camelid The end goal is to be able to use `hir::Type` directly. I also thought about storing it like this, but that seems a bit overkill. I'm just unsure for the moment which way is better. I'll try to get back to it.

**Comment by camelid** [comment]
> The end goal is to be able to use `hir::Type` directly.

Ah, that makes sense.

> I also thought about storing it like this, but that seems a bit overkill.

What do you mean by "overkill"?

> I'll try to get back to it.

👍

**Comment by GuillaumeGomez** [comment]
We already store the traits, so instead of having a second Hashmap, I think I simply create a struct wrapping the trait and the `is_spotlight` field.

**Comment by camelid** [comment]
> We already store the traits, so instead of having a second Hashmap, I think I'll simply create a struct wrapping the trait and the `is_spotlight` field.

It seems simpler to me to just have two hashmaps; we already have [`DocContext::auto_traits`](https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/core/struct.DocContext.html#structfield.auto_traits). But I'll wait until I see fully what you're proposing :)

**Comment by GuillaumeGomez** [comment]
Well, first getting #80883 to get merged and then we'll see. :p

**Comment by jyn514** [comment]
> That's actually a pretty good point. It forces us to run this check every time a function returns a "spotlighted" item. That's not great.

If you want to make it more efficient, you could calculate this once per item instead of once per function. Unless there's a noticable loss in performance, I don't think this needs to be cached, it looks relatively cheap and caching will increase the memory usage again.

> bool is only one byte, so I wouldn't think this would reduce memory usage that much.

It adds up when you have hundreds or thousands of items. Has this had a perf run yet?

**Comment by jyn514** [comment]
Oh wait, I didn't realize this is linear with the number of impls, ugh. Is there a query for this? That would get us caching automatically and more efficiently.

**Comment by camelid** [comment]
> Has this had a perf run yet?

Not to my knowledge.

**Comment by jyn514** [comment]
Oh right, this is waiting on https://github.com/rust-lang/rust/pull/80883 so we can get an accurate comparison.

**Comment by GuillaumeGomez** [comment]
Once #80883 is merged, I'll get back to this one. I have an idea on how to handle it (keeping the secret until I'm sure it's not a bad idea haha).

**Comment by GuillaumeGomez** [comment]
So the idea I had was to simply move the storage of this field inside the `Cache` in the `traits` hash map alongside the `clean` type. What do you think of this take?

**Comment by GuillaumeGomez** [comment]
Ignored the docblock I added.

**Comment by GuillaumeGomez** [comment]
ping @camelid @jyn514

**Comment by camelid** [comment]
I didn't notice this before because it was still marked as `S-waiting-on-author`.

File: src/librustdoc/core.rs
Comment (camelid) [inline src/librustdoc/core.rs:55]: I feel pretty strongly that a struct should be extracted here, for two reasons:

1. It's unclear what the `bool` represents — it could mean anything
2. It's very possible that we'll add more fields in the future, and we don't want to tack them on to the tuple

(This also applies for the other place `external_traits` is defined.)

**Review by camelid — CHANGES_REQUESTED** [review]

File: src/librustdoc/clean/utils.rs
Comment (camelid) [inline src/librustdoc/clean/utils.rs:469]: `list_contains_name` is a very generic function name. Maybe use a qualified path instead?

```suggestion
            && attr.meta_item_list().map_or(false, |l| rustc_attr::list_contains_name(&l, flag))
```

File: src/librustdoc/core.rs
Comment (GuillaumeGomez) [reply src/librustdoc/core.rs:55]: I hesitated for a while and thought it might be best to keep it like this. If I add a comment, is it fine for you or do you still prefer that I create a new type wrapping `clean::Trait` and the `bool`?

**Comment by jyn514** [comment]
Wow, that perf looks surprisingly good. I'd be interested to see how this scales to lots of blanket impls; I wish https://github.com/rust-lang/rustc-perf/pull/802 were working.

**Comment by GuillaumeGomez** [comment]
Nice work on rust-lang/rustc-perf#802 ! I also find the perf improvement surprisingly high. I expected that it'd be close to nothing, 0.2% is a great surprise! I'm just waiting for an answer from @camelid then I'll update the PR.

File: src/librustdoc/core.rs
Comment (camelid) [reply src/librustdoc/core.rs:55]: > thought it might be best to keep it like this

Can you explain your thinking a bit?

> If I add a comment, is it fine for you or do you still prefer that I create a new type wrapping `clean::Trait` and the `bool`?

I still prefer a new type :)

Many people likely won't read the comment. Besides, what's wrong with creating a new type? If for some reason it's a problem we can always change it to a tuple. Having a field name makes it much clearer what the `bool` is for, makes it much harder to introduce a bug by misunderstanding the meaning, and means people will not have to search to find docs in order to interact with it.

Comment (GuillaumeGomez) [reply src/librustdoc/core.rs:55]: Well... It's yet another type and we're trying to reduce that number. But, well, I don't mind making a new type so let's go for it!

**Comment by GuillaumeGomez** [comment]
Added the struct and updated the code as you suggested @camelid !

File: src/librustdoc/core.rs
Comment (jyn514) [reply src/librustdoc/core.rs]: If it helps, my motivation for removing types is not so much "more types are bad", but that we do a lot of shuffling around of the same data between different structures. Since this isn't copying the data, just refactoring out a type, I think it's a good change.

File: src/librustdoc/clean/types.rs
Comment (camelid) [inline src/librustdoc/clean/types.rs:75]: Why does this constructor exist? It negates some of the benefit of having a field name and I don't see any reason to have it.

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:75]: Ok, removing it then!

**Comment by GuillaumeGomez** [comment]
@camelid Removed the `new` method.

File: src/librustdoc/clean/utils.rs
Comment (camelid) [inline src/librustdoc/clean/utils.rs:457]: This seems a bit unclear. Maybe something along the lines of

```suggestion
/// Checks for the existence of a `doc(foo)` attribute. For example,
/// checks for the existence of `spotlight` in the attribute below:
```

would be clearer?

File: src/librustdoc/clean/types.rs
Comment (camelid) [inline src/librustdoc/clean/types.rs:60]: Is `external_traits` used only for traits accessible from outside the current crate? If so, does that mean `doc(spotlight)` will no longer work for private traits? I'm assuming it will still work because of `Cache.traits`, but I want to check to be absolutely sure :)

File: src/librustdoc/formats/cache.rs
Comment (camelid) [inline src/librustdoc/formats/cache.rs:258]: Not necessary to change now since this code was here before, but couldn't we simplify this code and avoid the `Entry` API since we aren't using the returned `&mut V`?

File: src/librustdoc/html/render/mod.rs
Comment (camelid) [inline src/librustdoc/html/render/mod.rs:3685]: Was `def_id_full` changed to `def_id` because there's no such thing as a primitive trait?

File: src/librustdoc/html/render/mod.rs
Comment (camelid) [inline src/librustdoc/html/render/mod.rs:3686]: Why was this changed from panicking to defaulting to `false`? Is this unrelated to the `is_spotlight` field removal?

File: src/librustdoc/clean/utils.rs
Comment (camelid) [inline src/librustdoc/clean/utils.rs:463]: nit:

```suggestion
/// This function has to exist because it runs on `hir::Attributes` whereas the other runs on
```

or

```suggestion
/// This function exists because it runs on `hir::Attributes` whereas the other runs on
```

Comment (camelid) [reply src/librustdoc/clean/utils.rs:463]: Might also be worth mentioning what the other function is, but not a big deal.

**Review by camelid — COMMENTED** [review]
Looks like this is almost ready! Left a few nits and questions.

Also, it looks like we have relatively little test coverage for `doc(spotlight)`. Do you want to add a test for an out-of-crate spotlighted trait to make sure this doesn't break those?

File: src/librustdoc/clean/types.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:60]: It doesn't impact the traits in the current crate if that's your question.

File: src/librustdoc/formats/cache.rs
Comment (GuillaumeGomez) [reply src/librustdoc/formats/cache.rs:258]: I added a FIXME.

File: src/librustdoc/html/render/mod.rs
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:3685]: Absolutely.

File: src/librustdoc/html/render/mod.rs
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:3686]: Both code do the same thing as far as I can tell. Just in more "convoluted" ways.

File: src/librustdoc/clean/utils.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/utils.rs:463]: Did it. :)

File: src/librustdoc/clean/utils.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/utils.rs:457]: Sounds good to me.

**Comment by GuillaumeGomez** [comment]
Fixed the typo, improved the doc and added a test to ensure that spotlighted traits from dependencies are taken into account as expected.

File: src/librustdoc/html/render/mod.rs
Comment (camelid) [reply src/librustdoc/html/render/mod.rs:3686]: I don't think so. Before it would panic if `d` did not exist in `cache.traits`. Now it just doesn't run the branch. The new behavior is probably better, but I want to make sure it's intentional.

File: src/librustdoc/clean/utils.rs
Comment (camelid) [inline src/librustdoc/clean/utils.rs:465]:
```suggestion
/// `clean::Attributes` method.
```

File: src/librustdoc/formats/cache.rs
Comment (camelid) [inline src/librustdoc/formats/cache.rs:252]:
```suggestion
            // FIXME: We don't use the returned `&mut` so better not use the `entry` API here.
```

File: src/test/rustdoc/spotlight-from-dependency.rs
Comment (camelid) [inline src/test/rustdoc/spotlight-from-dependency.rs:6]: nit: Might be clearer if this `@has` check were next to `pub fn new()` instead, to make it clear that the dialog is for the function.

**Review by camelid — COMMENTED** [review]
Three small nits; no need to fix, but if so that would be great!

File: src/test/rustdoc/spotlight-from-dependency.rs
Comment (GuillaumeGomez) [reply src/test/rustdoc/spotlight-from-dependency.rs]: Added it.

**Comment by GuillaumeGomez** [comment]
@camelid Updated!

File: src/test/rustdoc/spotlight-from-dependency.rs
Comment (camelid) [reply src/test/rustdoc/spotlight-from-dependency.rs]: Ah, I meant move the physical `@has` comment. But what you did is fine too :)

**Comment by camelid** [comment]
@jyn514 do you have any comments aside from the 👍 on https://github.com/rust-lang/rust/pull/80914#discussion_r575869229?

File: src/librustdoc/clean/utils.rs
Comment (jyn514) [inline src/librustdoc/clean/utils.rs:462]: It would be nice to make this a valid doc-test.
```suggestion
/// checks for the existence of `hidden` in the attribute below:
///
/// ```
/// #[doc(hidden)]
/// pub fn foo() {}
/// ```
```

File: src/librustdoc/formats/cache.rs
Comment (jyn514) [reply src/librustdoc/formats/cache.rs:259]: I disagree with the FIXME. Entry is more efficient because it only has to calculate the hash once (compared to `.contains`, which has to hash twice).

File: src/librustdoc/formats/cache.rs
Comment (jyn514) [inline src/librustdoc/formats/cache.rs:258]: You already have the attributes available as `item.attrs`, you don't need `get_attrs`. That should also let you get rid of the new `has_doc_flag` function since you'll be working on `clean::Attributes`.

File: src/librustdoc/clean/types.rs
Comment (camelid) [reply src/librustdoc/clean/types.rs:70]: The whole point of this PR is to remove the field from `Trait`, isn't it? I feel a bit confused as well.

Comment (jyn514) [reply src/librustdoc/clean/types.rs:70]: Oh right, that too. Why not calculate this on demand?

Comment (camelid) [reply src/librustdoc/clean/types.rs:70]: https://github.com/rust-lang/rust/pull/80914#pullrequestreview-578522364

Maybe we could use some sort of on-demand system with caching? Hmm... that makes me thinking of queries—can we define queries in rustdoc?

I wonder why this is having a positive perf impact given that it's still calculating ahead-of-time... I must be missing something.

File: src/librustdoc/formats/cache.rs
Comment (GuillaumeGomez) [reply src/librustdoc/formats/cache.rs:258]: Oh good catch! I'll remove this one. However, I still need this function because of `record_extern_trait` which, in certain contexts, doesn't have a `clean::Item` (like in `clean::utils::register_res`).

File: src/librustdoc/clean/types.rs
Comment (jyn514) [reply src/librustdoc/clean/types.rs:70]: Queries must be declared in rustc_middle, which means any types they use must be available that earlier. So we can't have a query that returns `clean::Trait`, but it's feasible to add one for `implement_doc_spotlight(DefId) -> bool`. That would need sign-off from a compiler-team member though.

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:70]: Should I add a FIXME so we can do it in another PR?

File: src/librustdoc/html/render/mod.rs
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/mod.rs:3685]: It is. :)

File: src/librustdoc/clean/types.rs
Comment (jyn514) [reply src/librustdoc/clean/types.rs:70]: Well, I'm not sure what the point of this PR is if you're still storing `is_spotlight` in memory. Why not calculate it everywhere it's used?

Comment (GuillaumeGomez) [reply src/librustdoc/clean/types.rs:70]: The point is to make `ItemKind` looks as close as possible to the compiler type so that it makes the switch simpler (which is why I make all these "small" PRs).

Comment (camelid) [reply src/librustdoc/clean/types.rs:70]: > Queries must be declared in rustc_middle, which means any types they use must be available that early. So we can't have a query that returns `clean::Trait`, but it's feasible to add one for `implement_doc_spotlight(DefId) -> bool`. That would need sign-off from a compiler-team member though.

Better yet, what about adding `has_doc_flag(DefId, Symbol) -> bool` so it's more general-purpose?

**Comment by GuillaumeGomez** [comment]
I think I fixed all comments. Ping @jyn514 @camelid

File: src/librustdoc/clean/inline.rs
Comment (jyn514) [inline src/librustdoc/clean/inline.rs:630]: nit: rustfmt really mangles this formatting, do you mind fixing it manually?
```suggestion
    let trait_ = clean::TraitWithExtraInfo {
        trait_,
        is_spotlight: clean::utils::has_doc_flag(cx.tcx.get_attrs(did), sym::spotlight),
    };
    cx.external_traits.borrow_mut().insert(did, trait_);
```

File: src/librustdoc/clean/types.rs
Comment (jyn514) [reply src/librustdoc/clean/types.rs:70]: > The point is to make ItemKind looks as close as possible to the compiler type so that it makes the switch simpler (which is why I make all these "small" PRs).

Could you add that to the PR description?

File: src/librustdoc/clean/utils.rs
Comment (jyn514) [inline src/librustdoc/clean/utils.rs:534]: nit: This is ambiguous to the reader (even though the compiler knows it's from `rustc_middle`) because there's also `clean::Attributes`.
```suggestion
crate fn has_doc_flag(attrs: ty::Attributes<'_>, flag: Symbol) -> bool {
```

**Review by jyn514 — COMMENTED** [review]
r=me with nits fixed and the description updated - I don't think this needs to mention https://github.com/rust-lang/rust/pull/80883 now that it's been merged.

File: src/librustdoc/clean/inline.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/inline.rs:630]: Sure.

File: src/librustdoc/clean/utils.rs
Comment (GuillaumeGomez) [reply src/librustdoc/clean/utils.rs]: Done as well.

**Comment by GuillaumeGomez** [comment]
> r=me with nits fixed and the description updated - I don't think this needs to mention #80883 now that it's been merged.

I updated the description but I'll keep the mention to the other PR. ;)

**Pull request merged**
